File: james-ultimate/kotlin-scanner/src/main/kotlin/com/emersa/james/kotlin/KotlinSecurityScanner.kt
package com.emersa.james.kotlin

import com.google.gson.Gson
import kotlinx.coroutines.*
import okhttp3.OkHttpClient
import okhttp3.Request
import java.net.InetSocketAddress
import java.net.Socket
import java.security.MessageDigest
import java.io.File
import java.time.Instant
import java.util.concurrent.TimeUnit

/**
 * Kotlin Security Scanner
 * Modern, coroutine-based security scanning with Kotlin features
 * 
 * Copyright © 2025 Emersa Ltd. All Rights Reserved.
 */

// Sealed class for scan results
sealed class ScanResult {
    data class Success<T>(val data: T, val duration: Long) : ScanResult()
    data class Failure(val error: String, val code: Int = -1) : ScanResult()
}

// Data classes for results
data class PortScanResult(
    val host: String,
    val port: Int,
    val isOpen: Boolean,
    val service: String?,
    val risk: String,
    val responseTime: Long
)

data class HashResult(
    val filePath: String,
    val md5: String,
    val sha1: String,
    val sha256: String,
    val sizeBytes: Long,
    val calculationTime: Long,
    val timestamp: String
)

data class ApiSecurityResult(
    val url: String,
    val method: String,
    val statusCode: Int,
    val headers: Map<String, String>,
    val vulnerabilities: List<String>,
    val riskScore: Int,
    val recommendations: List<String>
)

// Service detection map
object ServiceDetector {
    private val commonPorts = mapOf(
        21 to Pair("FTP", "high"),
        22 to Pair("SSH", "medium"),
        23 to Pair("Telnet", "critical"),
        25 to Pair("SMTP", "medium"),
        53 to Pair("DNS", "low"),
        80 to Pair("HTTP", "low"),
        110 to Pair("POP3", "medium"),
        143 to Pair("IMAP", "medium"),
        443 to Pair("HTTPS", "low"),
        445 to Pair("SMB", "high"),
        3306 to Pair("MySQL", "high"),
        3389 to Pair("RDP", "critical"),
        5432 to Pair("PostgreSQL", "high"),
        6379 to Pair("Redis", "high"),
        8080 to Pair("HTTP-Alt", "low"),
        27017 to Pair("MongoDB", "medium")
    )
    
    fun detectService(port: Int): Pair<String, String> {
        return commonPorts[port] ?: Pair("Unknown", "low")
    }
}

class KotlinSecurityScanner {
    private val gson = Gson()
    private val client = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(10, TimeUnit.SECONDS)
        .build()
    
    /**
     * Scan single port with coroutines (non-blocking)
     */
    suspend fun scanPort(host: String, port: Int, timeout: Int = 1000): PortScanResult = withContext(Dispatchers.IO) {
        val startTime = System.currentTimeMillis()
        var isOpen = false
        
        try {
            Socket().use { socket ->
                socket.connect(InetSocketAddress(host, port), timeout)
                isOpen = true
            }
        } catch (e: Exception) {
            // Port closed or timeout
        }
        
        val responseTime = System.currentTimeMillis() - startTime
        val (service, risk) = ServiceDetector.detectService(port)
        
        PortScanResult(
            host = host,
            port = port,
            isOpen = isOpen,
            service = if (isOpen) service else null,
            risk = risk,
            responseTime = responseTime
        )
    }
    
    /**
     * Parallel port scanning using coroutines
     */
    suspend fun scanPortRange(
        host: String,
        startPort: Int,
        endPort: Int,
        timeout: Int = 1000
    ): List<PortScanResult> = coroutineScope {
        val jobs = (startPort..endPort).map { port ->
            async {
                scanPort(host, port, timeout)
            }
        }
        
        jobs.awaitAll().filter { it.isOpen }
    }
    
    /**
     * Fast port scan (common ports only)
     */
    suspend fun scanCommonPorts(host: String): List<PortScanResult> = coroutineScope {
        val commonPorts = listOf(21, 22, 23, 25, 53, 80, 110, 143, 443, 445, 3306, 3389, 5432, 8080)
        
        val jobs = commonPorts.map { port ->
            async {
                scanPort(host, port, 500)
            }
        }
        
        jobs.awaitAll().filter { it.isOpen }
    }
    
    /**
     * Calculate file hashes (MD5, SHA-1, SHA-256)
     */
    fun calculateHash(filePath: String): ScanResult {
        val startTime = System.currentTimeMillis()
        
        return try {
            val file = File(filePath)
            if (!file.exists() || !file.isFile) {
                return ScanResult.Failure("File not found or not a file: $filePath")
            }
            
            val bytes = file.readBytes()
            
            val md5 = MessageDigest.getInstance("MD5")
                .digest(bytes)
                .joinToString("") { "%02x".format(it) }
            
            val sha1 = MessageDigest.getInstance("SHA-1")
                .digest(bytes)
                .joinToString("") { "%02x".format(it) }
            
            val sha256 = MessageDigest.getInstance("SHA-256")
                .digest(bytes)
                .joinToString("") { "%02x".format(it) }
            
            val duration = System.currentTimeMillis() - startTime
            
            val result = HashResult(
                filePath = filePath,
                md5 = md5,
                sha1 = sha1,
                sha256 = sha256,
                sizeBytes = file.length(),
                calculationTime = duration,
                timestamp = Instant.now().toString()
            )
            
            ScanResult.Success(result, duration)
        } catch (e: Exception) {
            ScanResult.Failure("Hash calculation failed: ${e.message}")
        }
    }
    
    /**
     * API Security Scanner
     */
    suspend fun scanApiSecurity(url: String, method: String = "GET"): ScanResult = withContext(Dispatchers.IO) {
        val startTime = System.currentTimeMillis()
        
        try {
            val request = Request.Builder()
                .url(url)
                .method(method, null)
                .build()
            
            val response = client.newCall(request).execute()
            
            val headers = response.headers.toMultimap()
                .mapValues { it.value.joinToString(", ") }
            
            val vulnerabilities = mutableListOf<String>()
            val recommendations = mutableListOf<String>()
            
            // Check security headers
            if (!headers.containsKey("Strict-Transport-Security")) {
                vulnerabilities.add("Missing HSTS header")
                recommendations.add("Add Strict-Transport-Security header")
            }
            
            if (!headers.containsKey("X-Content-Type-Options")) {
                vulnerabilities.add("Missing X-Content-Type-Options header")
                recommendations.add("Add X-Content-Type-Options: nosniff")
            }
            
            if (!headers.containsKey("X-Frame-Options")) {
                vulnerabilities.add("Missing X-Frame-Options header")
                recommendations.add("Add X-Frame-Options: DENY or SAMEORIGIN")
            }
            
            if (!headers.containsKey("Content-Security-Policy")) {
                vulnerabilities.add("Missing Content-Security-Policy")
                recommendations.add("Implement Content Security Policy")
            }
            
            if (headers["Server"]?.isNotEmpty() == true) {
                vulnerabilities.add("Server header exposes version information")
                recommendations.add("Remove or obfuscate Server header")
            }
            
            val riskScore = when (vulnerabilities.size) {
                0 -> 0
                1 -> 25
                2 -> 50
                3 -> 75
                else -> 100
            }
            
            val result = ApiSecurityResult(
                url = url,
                method = method,
                statusCode = response.code,
                headers = headers,
                vulnerabilities = vulnerabilities,
                riskScore = riskScore,
                recommendations = recommendations
            )
            
            val duration = System.currentTimeMillis() - startTime
            response.close()
            
            ScanResult.Success(result, duration)
        } catch (e: Exception) {
            ScanResult.Failure("API scan failed: ${e.message}")
        }
    }
    
    /**
     * Execute command from TypeScript bridge
     */
    fun execute(command: String, params: Map<String, String>): String {
        return runBlocking {
            when (command) {
                "port_scan" -> {
                    val host = params["host"] ?: return@runBlocking gson.toJson(mapOf("error" to "Missing host"))
                    val startPort = params["startPort"]?.toIntOrNull() ?: 1
                    val endPort = params["endPort"]?.toIntOrNull() ?: 1024
                    
                    val results = scanPortRange(host, startPort, endPort)
                    gson.toJson(mapOf(
                        "host" to host,
                        "startPort" to startPort,
                        "endPort" to endPort,
                        "openPorts" to results,
                        "timestamp" to Instant.now().toString()
                    ))
                }
                
                "port_scan_fast" -> {
                    val host = params["host"] ?: return@runBlocking gson.toJson(mapOf("error" to "Missing host"))
                    
                    val results = scanCommonPorts(host)
                    gson.toJson(mapOf(
                        "host" to host,
                        "openPorts" to results,
                        "timestamp" to Instant.now().toString()
                    ))
                }
                
                "hash_file" -> {
                    val filePath = params["filePath"] ?: return@runBlocking gson.toJson(mapOf("error" to "Missing filePath"))
                    
                    when (val result = calculateHash(filePath)) {
                        is ScanResult.Success<*> -> gson.toJson(result.data)
                        is ScanResult.Failure -> gson.toJson(mapOf("error" to result.error))
                    }
                }
                
                "api_scan" -> {
                    val url = params["url"] ?: return@runBlocking gson.toJson(mapOf("error" to "Missing url"))
                    val method = params["method"] ?: "GET"
                    
                    when (val result = scanApiSecurity(url, method)) {
                        is ScanResult.Success<*> -> gson.toJson(result.data)
                        is ScanResult.Failure -> gson.toJson(mapOf("error" to result.error))
                    }
                }
                
                else -> gson.toJson(mapOf("error" to "Unknown command: $command"))
            }
        }
    }
    
    /**
     * Get scanner information
     */
    fun getInfo(): String {
        return gson.toJson(mapOf(
            "name" to "Kotlin Security Scanner",
            "version" to "2.0.0",
            "language" to "Kotlin",
            "features" to listOf(
                "Coroutine-based parallel scanning",
                "Port scanning",
                "File hashing",
                "API security analysis",
                "Modern Kotlin features"
            ),
            "capabilities" to mapOf(
                "async" to true,
                "coroutines" to true,
                "parallel" to true
            )
        ))
    }
}

// CLI entry point
fun main(args: Array<String>) {
    val scanner = KotlinSecurityScanner()
    
    when {
        args.isEmpty() -> {
            println("Kotlin Security Scanner v2.0.0")
            println("Usage:")
            println("  kotlin-scanner scan <host> [start-port] [end-port]")
            println("  kotlin-scanner hash <file-path>")
            println("  kotlin-scanner api <url> [method]")
            println("  kotlin-scanner info")
        }
        
        args[0] == "info" -> {
            println(scanner.getInfo())
        }
        
        args[0] == "scan" && args.size >= 2 -> {
            val host = args[1]
            val startPort = args.getOrNull(2)?.toIntOrNull() ?: 1
            val endPort = args.getOrNull(3)?.toIntOrNull() ?: 1024
            
            runBlocking {
                println("Scanning $host ports $startPort-$endPort...")
                val results = scanner.scanPortRange(host, startPort, endPort)
                println("Found ${results.size} open ports:")
                results.forEach { result ->
                    println("  Port ${result.port}: ${result.service} (${result.risk} risk) - ${result.responseTime}ms")
                }
            }
        }
        
        args[0] == "hash" && args.size >= 2 -> {
            val filePath = args[1]
            when (val result = scanner.calculateHash(filePath)) {
                is ScanResult.Success<*> -> {
                    val hashResult = result.data as HashResult
                    println("Hash calculation for: ${hashResult.filePath}")
                    println("  MD5:    ${hashResult.md5}")
                    println("  SHA-1:  ${hashResult.sha1}")
                    println("  SHA-256: ${hashResult.sha256}")
                    println("  Size:   ${hashResult.sizeBytes} bytes")
                    println("  Time:   ${hashResult.calculationTime}ms")
                }
                is ScanResult.Failure -> println("Error: ${result.error}")
            }
        }
        
        args[0] == "api" && args.size >= 2 -> {
            val url = args[1]
            val method = args.getOrNull(2) ?: "GET"
            
            runBlocking {
                println("Scanning API: $url")
                when (val result = scanner.scanApiSecurity(url, method)) {
                    is ScanResult.Success<*> -> {
                        val apiResult = result.data as ApiSecurityResult
                        println("Status Code: ${apiResult.statusCode}")
                        println("Risk Score: ${apiResult.riskScore}/100")
                        println("\nVulnerabilities:")
                        apiResult.vulnerabilities.forEach { println("  - $it") }
                        println("\nRecommendations:")
                        apiResult.recommendations.forEach { println("  - $it") }
                    }
                    is ScanResult.Failure -> println("Error: ${result.error}")
                }
            }
        }
        
        else -> {
            println("Unknown command. Use 'kotlin-scanner' without arguments for help.")
        }
    }
}

File: james-ultimate/src/kotlin-bridge/KotlinSecurityScanner.ts
/**
 * Kotlin Security Scanner Bridge
 * 
 * Provides TypeScript/Node.js interface to Kotlin security scanner.
 * Uses child_process to communicate with Kotlin JAR.
 * 
 * Copyright © 2025 Emersa Ltd. All Rights Reserved.
 */

import * as path from 'path';
import { spawn, exec } from 'child_process';
import { EventEmitter } from 'events';
import { promisify } from 'util';

const execAsync = promisify(exec);

export interface KotlinPortScanResult {
    host: string;
    port: number;
    isOpen: boolean;
    service: string | null;
    risk: string;
    responseTime: number;
}

export interface KotlinHashResult {
    filePath: string;
    md5: string;
    sha1: string;
    sha256: string;
    sizeBytes: number;
    calculationTime: number;
    timestamp: string;
}

export interface KotlinApiSecurityResult {
    url: string;
    method: string;
    statusCode: number;
    headers: Record<string, string>;
    vulnerabilities: string[];
    riskScore: number;
    recommendations: string[];
}

export class KotlinSecurityScanner extends EventEmitter {
    private jarPath: string;
    private initialized: boolean = false;
    
    constructor() {
        super();
        this.jarPath = path.join(__dirname, '../../kotlin-scanner/build/libs/kotlin-scanner-2.0.0.jar');
    }
    
    /**
     * Initialize Kotlin scanner (check if JAR exists)
     */
    async initialize(): Promise<void> {
        if (this.initialized) {
            return;
        }
        
        try {
            const fs = require('fs');
            if (!fs.existsSync(this.jarPath)) {
                throw new Error(`Kotlin scanner JAR not found at ${this.jarPath}. Run 'npm run build:kotlin'`);
            }
            
            // Test if Java is available
            await execAsync('java -version');
            
            this.initialized = true;
            this.emit('initialized');
            console.log('[KotlinBridge] Security scanner initialized successfully');
            
        } catch (error) {
            console.error('[KotlinBridge] Failed to initialize:', error);
            throw new Error(`Kotlin scanner initialization failed: ${error}`);
        }
    }
    
    /**
     * Ensure scanner is initialized
     */
    private async ensureInitialized(): Promise<void> {
        if (!this.initialized) {
            await this.initialize();
        }
    }
    
    /**
     * Execute Kotlin scanner command
     */
    private async executeCommand(args: string[]): Promise<string> {
        await this.ensureInitialized();
        
        return new Promise((resolve, reject) => {
            const java = spawn('java', ['-jar', this.jarPath, ...args]);
            
            let stdout = '';
            let stderr = '';
            
            java.stdout.on('data', (data) => {
                stdout += data.toString();
            });
            
            java.stderr.on('data', (data) => {
                stderr += data.toString();
            });
            
            java.on('close', (code) => {
                if (code !== 0) {
                    reject(new Error(`Kotlin scanner exited with code ${code}: ${stderr}`));
                } else {
                    resolve(stdout.trim());
                }
            });
            
            java.on('error', (error) => {
                reject(error);
            });
        });
    }
    
    /**
     * Execute via JSON API
     */
    private async execute(command: string, params: Record<string, string>): Promise<any> {
        await this.ensureInitialized();
        
        return new Promise((resolve, reject) => {
            // For JSON API, we'll need to add a JSON mode to the Kotlin scanner
            // For now, use command-line mode
            const args = [command];
            Object.entries(params).forEach(([key, value]) => {
                args.push(value);
            });
            
            this.executeCommand(args)
                .then(output => {
                    try {
                        // Try to parse as JSON
                        const result = JSON.parse(output);
                        resolve(result);
                    } catch (e) {
                        // If not JSON, return as is
                        resolve({ output });
                    }
                })
                .catch(reject);
        });
    }
    
    /**
     * Port scan - scan range of ports
     */
    async portScan(host: string, startPort: number, endPort: number): Promise<any> {
        console.log(`[KotlinBridge] Port scanning ${host}:${startPort}-${endPort}`);
        
        try {
            const output = await this.executeCommand(['scan', host, String(startPort), String(endPort)]);
            // Parse the output and format it
            return {
                host,
                startPort,
                endPort,
                output,
                timestamp: new Date().toISOString()
            };
        } catch (error) {
            throw new Error(`Port scan failed: ${error}`);
        }
    }
    
    /**
     * Fast port scan - scan common ports only
     */
    async portScanFast(host: string): Promise<any> {
        console.log(`[KotlinBridge] Fast port scanning ${host}`);
        
        return this.portScan(host, 1, 1024);
    }
    
    /**
     * Calculate file hash
     */
    async hashFile(filePath: string): Promise<KotlinHashResult> {
        console.log(`[KotlinBridge] Calculating hash for ${filePath}`);
        
        try {
            const output = await this.executeCommand(['hash', filePath]);
            const result = JSON.parse(output);
            return result as KotlinHashResult;
        } catch (error) {
            throw new Error(`Hash calculation failed: ${error}`);
        }
    }
    
    /**
     * Scan API security
     */
    async scanApiSecurity(url: string, method: string = 'GET'): Promise<KotlinApiSecurityResult> {
        console.log(`[KotlinBridge] Scanning API ${url}`);
        
        try {
            const output = await this.executeCommand(['api', url, method]);
            const result = JSON.parse(output);
            return result as KotlinApiSecurityResult;
        } catch (error) {
            throw new Error(`API scan failed: ${error}`);
        }
    }
    
    /**
     * Get scanner information
     */
    async getInfo(): Promise<any> {
        await this.ensureInitialized();
        
        try {
            const output = await this.executeCommand(['info']);
            return JSON.parse(output);
        } catch (error) {
            throw new Error(`Failed to get info: ${error}`);
        }
    }
    
    /**
     * Check if Kotlin scanner is available
     */
    static isAvailable(): boolean {
        try {
            const fs = require('fs');
            const jarPath = path.join(__dirname, '../../kotlin-scanner/build/libs/kotlin-scanner-2.0.0.jar');
            return fs.existsSync(jarPath);
        } catch {
            return false;
        }
    }
    
    /**
     * Shutdown and cleanup
     */
    async shutdown(): Promise<void> {
        if (this.initialized) {
            this.initialized = false;
            this.emit('shutdown');
            console.log('[KotlinBridge] Scanner shutdown');
        }
    }
}

// Singleton instance
let scannerInstance: KotlinSecurityScanner | null = null;

/**
 * Get or create scanner instance
 */
export function getKotlinScanner(): KotlinSecurityScanner {
    if (!scannerInstance) {
        scannerInstance = new KotlinSecurityScanner();
    }
    return scannerInstance;
}

/**
 * Check if Kotlin scanning is available
 */
export function isKotlinAvailable(): boolean {
    return KotlinSecurityScanner.isAvailable();
}

export default KotlinSecurityScanner;
